File: VB/S131084/Form1.vb
Imports DevExpress.XtraCharts
Imports DevExpress.XtraGrid.Views.Base
Imports DevExpress.XtraGrid.Views.Layout
Imports DevExpress.XtraGrid.Views.Layout.ViewInfo
Imports System
Imports System.Collections.Generic
Imports System.Data
Imports System.Drawing
Imports System.Windows.Forms

Namespace S131084
    Partial Public Class Form1
        Inherits Form

        Private chart As ChartControl
        Private dataSet As DataSet
        Private unboundData As New Dictionary(Of Integer, Bitmap)()

        Public Sub New()
            InitializeComponent()
            InitializeChart()
            InitializeSource()
        End Sub

        Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs) Handles Me.Load
            productsBindingSource.DataSource = dataSet.Tables("Products")
            personsBindingSource.DataSource = dataSet.Tables("Persons")
        End Sub

        Private Function GetProductsDataTable() As DataTable
            Dim table As New DataTable()
            table.TableName = "Products"
            table.Columns.Add(New DataColumn("ProductName", GetType(String)))
            table.Columns.Add(New DataColumn("UnitPrice", GetType(Double)))
            table.Columns.Add(New DataColumn("ID", GetType(Integer)))
            For i As Integer = 0 To 999
                table.Rows.Add("Product " & i, 3000 + i * 298.55D, i Mod 200)
            Next i
            Return table
        End Function
        Private Function GetPersonDataTable() As DataTable
            Dim table As New DataTable()
            table.TableName = "Persons"
            table.Columns.Add(New DataColumn("FirstName", GetType(String)))
            table.Columns.Add(New DataColumn("SecondName", GetType(String)))
            table.Columns.Add(New DataColumn("Image", GetType(Image)))
            table.Columns.Add(New DataColumn("ID", GetType(Integer)))
            For i As Integer = 0 To 199
                table.Rows.Add("FirstName " & i, "SecondName " & i, My.Resources.picture_post_card, i)
            Next i
            Return table
        End Function

        Private Sub InitializeChart()
            chart = New ChartControl()
            Dim series1 As New Series("series1", ViewType.Bar)
            series1.ArgumentDataMember = "ProductName"
            series1.ValueDataMembers.AddRange(New String() { "UnitPrice" })
            series1.ShowInLegend = False
            Dim series2 As New Series("series2", ViewType.Point)
            series2.ArgumentDataMember = "ProductName"
            series2.ValueDataMembers.AddRange(New String() { "UnitPrice" })
            series2.ShowInLegend = False
            series2.Label.Visible = False
            chart.Series.AddRange(New Series() { series1, series2 })
            CType(chart.Diagram, XYDiagram).AxisX.Label.Angle = 20
        End Sub
        Private Sub InitializeSource()
            dataSet = New DataSet()
            dataSet.Tables.Add(GetProductsDataTable())
            dataSet.Tables.Add(GetPersonDataTable())
            Dim keyColumn As DataColumn = dataSet.Tables("Persons").Columns("ID")
            Dim foreignKeyColumn As DataColumn = dataSet.Tables("Products").Columns("ID")
            dataSet.Relations.Add("PersonsProducts", keyColumn, foreignKeyColumn)
        End Sub

        Private Sub layoutView1_CustomUnboundColumnData(ByVal sender As Object, ByVal e As CustomColumnDataEventArgs) Handles layoutView1.CustomUnboundColumnData
            If e.Column Is colProducts Then
                If e.IsGetData Then
                    Dim personId As Integer = CInt((dataSet.Tables("Persons").Rows(e.ListSourceRowIndex)("ID")))
                    Dim img As Bitmap = Nothing
                    If unboundData.ContainsKey(personId) Then
                        img = unboundData(personId)
                    Else
                        Dim dataView As DataView = dataSet.Tables("Products").DefaultView
                        dataView.RowFilter = String.Format("ID = {0}", personId)
                        For Each s As Series In chart.Series
                            s.DataSource = dataView
                        Next s
                        chart.RefreshData()
                        Dim view As LayoutView = DirectCast(sender, LayoutView)
                        Dim viewInfo As LayoutViewInfo = CType(view.GetViewInfo(), LayoutViewInfo)
                        Dim chartSize As Size = Size.Empty
                        For Each field As LayoutViewField In view.TemplateCard.Items
                            If field.FieldName = "colProducts" Then
                                chartSize = field.Size
                                Exit For
                            End If
                        Next field
                        chart.Size = chartSize
                        img = New Bitmap(chartSize.Width, chartSize.Height)
                        chart.DrawToBitmap(img, New Rectangle(Point.Empty, chartSize))
                        unboundData.Add(personId, img)
                    End If
                    e.Value = img
                End If
            End If
        End Sub
    End Class
End Namespace

File: CS/S131084/Form1.cs
using DevExpress.XtraCharts;
using DevExpress.XtraGrid.Views.Base;
using DevExpress.XtraGrid.Views.Layout;
using DevExpress.XtraGrid.Views.Layout.ViewInfo;
using S131084.Properties;
using System;
using System.Collections.Generic;
using System.Data;
using System.Drawing;
using System.Windows.Forms;

namespace S131084
{
    public partial class Form1 : Form
    {
        private ChartControl chart;
        DataSet dataSet;
        private Dictionary<int, Bitmap> unboundData = new Dictionary<int, Bitmap>();

        public Form1()
        {
            InitializeComponent();
            InitializeChart();
            InitializeSource();
        }

        private void Form1_Load(object sender, EventArgs e)
        {
            productsBindingSource.DataSource = dataSet.Tables["Products"];
            personsBindingSource.DataSource = dataSet.Tables["Persons"];
        }

        DataTable GetProductsDataTable()
        {
            DataTable table = new DataTable();
            table.TableName = "Products";
            table.Columns.Add(new DataColumn("ProductName", typeof(string)));
            table.Columns.Add(new DataColumn("UnitPrice", typeof(double)));
            table.Columns.Add(new DataColumn("ID", typeof(int)));
            for (int i = 0; i < 1000; i++)
                table.Rows.Add("Product " + i, 3000 + i * 298.55M, i % 200);
            return table;
        }
        DataTable GetPersonDataTable()
        {
            DataTable table = new DataTable();
            table.TableName = "Persons";
            table.Columns.Add(new DataColumn("FirstName", typeof(string)));
            table.Columns.Add(new DataColumn("SecondName", typeof(string)));
            table.Columns.Add(new DataColumn("Image", typeof(Image)));
            table.Columns.Add(new DataColumn("ID", typeof(int)));
            for (int i = 0; i < 200; i++)
                table.Rows.Add("FirstName " + i, "SecondName " + i, Resources.picture_post_card, i);
            return table;
        }

        private void InitializeChart()
        {
            chart = new ChartControl();
            Series series1 = new Series("series1", ViewType.Bar);
            series1.ArgumentDataMember = "ProductName";
            series1.ValueDataMembers.AddRange(new string[] { "UnitPrice" });
            series1.ShowInLegend = false;
            Series series2 = new Series("series2", ViewType.Point);
            series2.ArgumentDataMember = "ProductName";
            series2.ValueDataMembers.AddRange(new string[] { "UnitPrice" });
            series2.ShowInLegend = false;
            series2.Label.Visible = false;
            chart.Series.AddRange(new Series[] { series1, series2 });
            ((XYDiagram)chart.Diagram).AxisX.Label.Angle = 20;
        }
        private void InitializeSource()
        {
            dataSet = new DataSet();
            dataSet.Tables.Add(GetProductsDataTable());
            dataSet.Tables.Add(GetPersonDataTable());
            DataColumn keyColumn = dataSet.Tables["Persons"].Columns["ID"];
            DataColumn foreignKeyColumn = dataSet.Tables["Products"].Columns["ID"];
            dataSet.Relations.Add("PersonsProducts", keyColumn, foreignKeyColumn);
        }

        private void layoutView1_CustomUnboundColumnData(object sender, CustomColumnDataEventArgs e)
        {
            if (e.Column == colProducts)
            {
                if (e.IsGetData)
                {
                    int personId = (int)dataSet.Tables["Persons"].Rows[e.ListSourceRowIndex]["ID"];
                    Bitmap img = null;
                    if (unboundData.ContainsKey(personId))
                        img = unboundData[personId];
                    else
                    {
                        DataView dataView = dataSet.Tables["Products"].DefaultView;
                        dataView.RowFilter = string.Format("ID = {0}", personId);
                        foreach (Series s in chart.Series)
                            s.DataSource = dataView;
                        chart.RefreshData();
                        LayoutView view = (LayoutView)sender;
                        LayoutViewInfo viewInfo = (LayoutViewInfo)view.GetViewInfo();
                        Size chartSize = Size.Empty;
                        foreach (LayoutViewField field in view.TemplateCard.Items)
                            if (field.FieldName == "colProducts")
                            {
                                chartSize = field.Size;
                                break;
                            }
                        chart.Size = chartSize;
                        img = new Bitmap(chartSize.Width, chartSize.Height);
                        chart.DrawToBitmap(img, new Rectangle(Point.Empty, chartSize));
                        unboundData.Add(personId, img);
                    }
                    e.Value = img;
                }
            }
        }
    }
}